### PR TITLE
pubsub/all: add an Message.BeforeSend As hook

### LINKED ISSUE #1907: pubsub: add an As hook for sending a message
We have a `Message.As` for accessing the provider's type for a received message, but no way for a user to set provider-specific things as a message is being sent.

### PR DESCRIPTION
**Pull request by vangent**
Fixes #1907.

### CODE REVIEW

File: pubsub/driver/driver.go [hunk 1]
```diff
@@ -52,6 +52,15 @@ type Message struct {
 	// see Topic.As for more details.
 	// AsFunc must be populated on messages returned from ReceiveBatch.
 	AsFunc func(interface{}) bool
+
+	// BeforeSend is a callback used when sending a message. It should remain
+	// nil on messages returned from ReceiveBatch.
+	//
+	// The callback must called exactly once, before the message is sent.
```
Comment (jba) [inline pubsub/driver/driver.go]: must => must be

File: pubsub/azuresb/azuresb.go [hunk 2]
```diff
@@ -269,6 +269,18 @@ func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
 	for k, v := range dm.Metadata {
 		sbms.Set(k, v)
 	}
+	if dm.BeforeSend != nil {
```
Comment (jba) [inline pubsub/azuresb/azuresb.go]: Missing "As" doc comment at top of file?

File: pubsub/driver/driver.go (hunk 1 above)
Comment (vangent) [reply pubsub/driver/driver.go]: Done.

File: pubsub/azuresb/azuresb.go (hunk 2 above)
Comment (vangent) [reply pubsub/azuresb/azuresb.go]: Done.

**Pull request merged**